🪟 🔧 UI Table migration to v8

YatsukBogdan1 opened the pull request:
## What
Closes #17941 

## How
Create a new component (`NextTable` temporary name) to use instead of `Table`, and migrate all `Table` occurrences to use new one

krishnaglick commented on airbyte-webapp/src/components/ui/NextTable/NextTable.tsx:
```diff
@@ -1,43 +1,87 @@
+    style.paddingLeft = `${meta?.customPadding?.left}px`;
+  }
+  if (meta?.customPadding?.right) {
+    style.paddingRight = `${meta?.customPadding?.right}px`;
```
Nit: You don't need to `?` within the if block.

krishnaglick commented on airbyte-webapp/src/components/ui/NextTable/NextTable.tsx:
```diff
@@ -1,43 +1,87 @@
-            ))}
+            {headerGroup.headers.map((header) => {
+              const { meta } = header.column.columnDef;
+              const style: CSSProperties = getStyleFromColumnMeta<TData<T>>(meta as ColumnMeta<T, unknown>);
```
Nit: Don't need CSSProperties here

krishnaglick commented on airbyte-webapp/src/components/ui/NextTable/NextTable.tsx:
```diff
@@ -1,43 +1,87 @@
+              return (
+                <th
+                  colSpan={header.colSpan}
+                  className={classNames(styles.th, meta?.thClassName, {
```
Does this filter out undefined or will it apply "undefined" as a class name?

YatsukBogdan1 replied:
just checked this. seems like `classNames` filters out `undefined` value, so it is not going into DOM

krishnaglick commented on airbyte-webapp/src/packages/cloud/views/credits/CreditsPage/components/UsagePerConnectionTable.tsx:
```diff
@@ -44,7 +45,7 @@ const UsagePerConnectionTable: React.FC<UsagePerConnectionTableProps> = ({ credi
   const { destinationDefinitions } = useDestinationDefinitionList();
 
-  const creditConsumptionWithPercent = React.useMemo(() => {
+  const creditConsumptionWithPercent = React.useMemo((): FullTableProps[] => {
```
Nit:
```suggestion
  const creditConsumptionWithPercent = React.useMemo<FullTableProps[]>(() => {
```

krishnaglick commented on airbyte-webapp/src/packages/cloud/views/credits/CreditsPage/components/UsagePerConnectionTable.tsx:
```diff
@@ -102,14 +104,14 @@ const UsagePerConnectionTable: React.FC<UsagePerConnectionTableProps> = ({ credi
 
   const sortingData = React.useMemo(
-    () => creditConsumptionWithPercent.sort(sortData),
+    (): FullTableProps[] => creditConsumptionWithPercent.sort(sortData),
```
Nit: Move above to `useMemo<FullTableProps[]>`

krishnaglick commented on airbyte-webapp/src/packages/cloud/views/credits/CreditsPage/components/UsagePerConnectionTable.tsx:
```diff
@@ -102,14 +104,14 @@ const UsagePerConnectionTable: React.FC<UsagePerConnectionTableProps> = ({ credi
 
   const columns = React.useMemo(
-    () => [
+    (): Array<ColumnDef<FullTableProps>> => [
```
Nit: Same here

krishnaglick commented on airbyte-webapp/src/packages/cloud/views/credits/CreditsPage/components/UsagePerConnectionTable.tsx:
```diff
@@ -139,31 +145,45 @@ const UsagePerConnectionTable: React.FC<UsagePerConnectionTableProps> = ({ credi
   return (
     <Content>
-      <Table columns={columns} data={sortingData} light />
+      <NextTable<FullTableProps> columns={columns} data={sortingData} light />
```
I did some testing to remove the @ts-ignore but ended up with an odd error here:
![Screenshot 2023-01-09 at 1 54 45 PM](https://user-images.githubusercontent.com/4967694/211385750-d60e33d3-5a6d-4866-8410-51adefcbcd44.png)
Seems like something can be improved in the type chain.

krishnaglick replied:
[Here](https://github.com/airbytehq/airbyte/blob/2713fa86192a20ed0c3964be47fbfd8669fda331/airbyte-webapp/src/components/ui/NextTable/NextTable.tsx#L14)
`columns: Array<ColumnDef<TData<T>, string>>;` seems to fix it.

dizel852 commented on airbyte-webapp/src/components/ui/NextTable/NextTable.tsx:
```diff
@@ -1,43 +1,87 @@
-export const NextTable = <TData,>({ columns, data, onClickRow, className }: PropsWithChildren<TableProps<TData>>) => {
+const getStyleFromColumnMeta = <T,>(meta: ColumnMeta<T, unknown>): CSSProperties => {
+  const style: CSSProperties = {};
+  if (meta?.customWidth) {
```
any reason why we introduced the style props?
We can pass the desired style via css in the meta

dizel852 commented on airbyte-webapp/src/types/react-table.d.ts:
```diff
@@ -7,7 +7,14 @@ import "@tanstack/react-table";
   interface ColumnMeta {
-    thClassName?: string;
+    collapse?: boolean;
+    customPadding?: {
```
I'd prefer to avoid adding style props here since after some time we will get the same result as we have in the current implementation.

krishnaglick commented on airbyte-webapp/src/components/ui/NextTable/NextTable.tsx:
```diff
@@ -1,43 +1,72 @@
+export interface TableProps<T> {
   className?: string;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  columns: Array<ColumnDef<TData<T>, any>>;
```
Don't love this any but unsure if we can do anything about it. It's fine, just would like it if it was unnecessary.

YatsukBogdan1 replied:
I left a comment in the `NextTable` component. We tried yesterday with @dizel852 to get rid of `any` here, but we didn't find a way. It also seems to be type safely, since the `useReactTable` `options.columns` itself expects `Array<ColumnDef<T, any>>`

dizel852 replied:
Confirm, double checked that

krishnaglick reviewed (commented):
Nothing stands out to me w/ the current state of things!

dizel852 commented on airbyte-webapp/src/components/ui/NextTable/NextTable.module.scss:
```diff
@@ -18,23 +23,48 @@
   &:last-child {
-    padding-left: vars.$spacing-lg;
-    border-radius: 0 vars.$border-radius-md 0 0;
+    padding-left: 15px;
+    border-radius: 0 10px 0 0;
```
Please fix with css variables

dizel852 commented on airbyte-webapp/src/components/ui/NextTable/NextTable.module.scss:
```diff
@@ -18,23 +23,48 @@
-    padding-left: calc(vars.$spacing-2xl + vars.$spacing-sm);
-    border-radius: vars.$border-radius-md 0 0;
+    padding-left: vars.$spacing-2xl;
+    border-radius: 10px 0 0;
```
css variables was removed

dizel852 commented on airbyte-webapp/src/components/EntityTable/ConnectionTable.tsx:
```diff
@@ -25,6 +27,16 @@ interface IProps {
   onSync: (id: string) => void;
 }
 
+type ColumnDefs = [
+  ColumnDef<ITableDataItem, string>,
+  ColumnDef<ITableDataItem, string>,
+  ColumnDef<ITableDataItem, string>,
+  ColumnDef<ITableDataItem, ConnectionScheduleData>,
+  ColumnDef<ITableDataItem, number>,
+  ColumnDef<ITableDataItem, boolean>,
+  ColumnDef<ITableDataItem, string>
+];
+
```
Please, use `columnHelper()` from `@tanstack/react-table` so we don't need to redefine types by ourselves.
Doc: https://tanstack.com/table/v8/docs/guide/column-defs#column-helpers

dizel852 commented on airbyte-webapp/src/components/ui/NextTable/NextTable.module.scss:
```diff
@@ -18,23 +23,48 @@
   line-height: 12px;
   font-weight: 600;
+  width: auto;
+  padding: 9px 13px 10px;
```
css variables

dizel852 commented on airbyte-webapp/src/components/ui/NextTable/NextTable.tsx:
```diff
@@ -1,43 +1,72 @@
+
+export interface TableProps<T> {
   className?: string;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
I'm pretty sure we don't need to provide `any` type here since the goal of refactoring was to make out base table more strictly typed

dizel852 replied:
Double-checked - we can use `any` here since under the hood react-table uses the same interface

dizel852 commented on airbyte-webapp/src/components/EntityTable/ConnectionTable.tsx:
```diff
@@ -146,39 +173,43 @@ const ConnectionTable: React.FC<IProps> = ({ data, entity, onClickRow, onSync })
   );
 
-  return <Table columns={columns} data={sortingData} onClickRow={onClickRow} erroredRows />;
+  return <NextTable columns={columns} data={sortingData} onClickRow={onClickRow} erroredRows />;
```
`erroredRows` - I don't the this in props, how do we get it?

dizel852 commented on airbyte-webapp/src/types/react-table.d.ts:
```diff
@@ -7,7 +7,9 @@ import "@tanstack/react-table";
   interface ColumnMeta {
-    thClassName?: string;
+    collapse?: boolean;
+    headerHighlighted?: boolean;
```
I'm not sure I understand what is `headerHighlighted`. 
Can we style instead?

tealjulia replied:
Agree, this should probably be styled using a classname.
I have a separate Issue that I created to clean up the styling on the tables (header alignment, header highlighting, etc.) https://github.com/airbytehq/airbyte/issues/21941

dizel852 commented on airbyte-webapp/src/packages/cloud/views/credits/CreditsPage/components/UsagePerConnectionTable.tsx:
```diff
@@ -139,31 +151,37 @@ const UsagePerConnectionTable: React.FC<UsagePerConnectionTableProps> = ({ credi
+        },
+        cell: (props) => <UsageCell percent={props.cell.getValue()} />,
       },
       // TODO: Replace to Grow column
```
Is this comment relevant?

dizel852 commented on airbyte-webapp/src/components/ui/NextTable/NextTable.module.scss:
```diff
@@ -18,23 +23,48 @@
+    font-weight: 400;
+  }
+
+  &.collapse {
+    /* stylelint-disable-next-line number-max-precision */
+    width: 0.0000000001%;
+  }
```
Just curious, this sounds to be like a row collapse.
Is this style just set minimum width?

dizel852 reviewed (changes requested):
Did a review, found some issues that need to be addressed:
1. We need to be really carefully updating base table styles, since it affects all other tables. Streams details panel is changed now:
Before changes:
<img width="1232" alt="Screenshot at Jan 20 17-39-24" src="https://user-images.githubusercontent.com/20929344/213742379-ba7da706-c70f-455f-8141-b4860e5225d3.png">

After changes:
<img width="1233" alt="Screenshot at Jan 20 17-41-55" src="https://user-images.githubusercontent.com/20929344/213742455-e078044e-7f58-4cbf-9485-2110be2d16cf.png">

2. Some props, like `erroredRows` and `collapse`, really sounds to me like something table-specific(not all tables can have errors). But we can extend the base component to change the `<tr/>` background color if we need. Except "light", since we use 2 styles of table header(blue and white)

3. Base table should be as strictly typed as as is can be. This will save us a lot of time in future

YatsukBogdan1 commented:
Fixed most of the comments, but there is still some questions I need answers for. I think @krishnaglick or @edmundito can give an answer
- do we need to keep `erroredRows` table property? seems like it should color rows which have `error: true` in red, but yesterday we were looking with @dizel852 into tables that used this property, and seems there is no actual results with this styling
- do we need to keep `headerHighlighted` inside column `meta`? seems this is not table specific feature, and can be done through direct styling of header
- I was trying to change all usages of magic numbers inside SCSS to variables, but some of values was specific, so I used the  closest values from SCSS variables; is this ok for us, or should we play around to make all spacing pixel perfect, as it was in original `Table` component?
- there is comment inside `src/packages/cloud/views/credits/CreditsPage/components/UsagePerConnectionTable.tsx` is it still relevant for us? and how we should handle it?
```
// TODO: Replace to Grow column
columnHelper.accessor("connectionId", {
  header: "",
  cell: () => <div />,
  meta: {
    thClassName: styles.thConnectionId,
  },
}),
```
Also:
- we investigated with @dizel852 on `react-table` `columns` typization and seems safely to have `NextTable` property `columns: Array<ColumnDef<TData<T>, any>>;` with `any` since `useReactTable` `options.columns` hook itself expects `Array<ColumnDef<T, any>>`

tealjulia commented on airbyte-webapp/src/components/ui/NextTable/NextTable.tsx:
```diff
@@ -1,43 +1,72 @@
+            className={classNames(styles.tr, {
+              // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+              // @ts-ignore
+              [styles.withError]: erroredRows && !!row.original.error,
```
I can't find a code path where we would need/want this.  I may be wrong here... but I _think_ this may be removable.

dizel852 replied:
+1 agree

tealjulia commented on airbyte-webapp/src/components/EntityTable/components/FrequencyCell.tsx:
```diff
@@ -27,7 +27,7 @@ const FrequencyCell: React.FC<FrequencyCellProps> = ({ value, enabled, scheduleT
     <Content enabled={enabled}>
       <FormattedMessage
         id={`frequency.${value ? value.basicSchedule?.timeUnit : "manual"}`}
```
```suggestion
        id={`frequency.${value?.basicSchedule?.timeUnit ?? "manual"}`}
```
more consistent since we're optional chaining below

tealjulia commented on airbyte-webapp/src/components/EntityTable/components/LastSyncCell.tsx:
```diff
@@ -7,7 +7,7 @@ const Content = styled.div<{ enabled?: boolean }>`
 
 interface IProps {
-  timeInSecond: number;
+  timeInSecond?: number | null;
```
nit: `timeInSeconds`

tealjulia commented on airbyte-webapp/src/components/EntityTable/components/LastSyncCell.tsx:
```diff
@@ -7,7 +7,7 @@ const Content = styled.div<{ enabled?: boolean }>`
 `;
 
 interface IProps {
```
nit: let's name the prop instead of IProps

tealjulia reviewed (changes requested):
Looks like we lost sortability on the credit usage table (credits page on cloud mode)

dizel852 commented on airbyte-webapp/src/components/ui/NextTable/NextTable.tsx:
```diff
@@ -1,43 +1,72 @@
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  columns: Array<ColumnDef<T, any>>;
+  data: T[];
+  erroredRows?: boolean;
```
I guess it came from an old design(probably) - For the case when the connection sync failed and we needed to color it red. And I also stand for removing it, since it doesn't make any effect.

tealjulia commented:
> there is comment inside src/packages/cloud/views/credits/CreditsPage/components/UsagePerConnectionTable.tsx is it still relevant for us? and how we should handle it?

@YatsukBogdan1 I'm not sure you got an answer to this one.  Compose team is implementing a full redesign of this table.  It should be ok to get rid of for now.  The main thing on this page is what I mentioned above (table no longer sorts).

dizel852 commented on airbyte-webapp/src/components/EntityTable/ConnectionTable.tsx:
```diff
@@ -16,12 +16,14 @@ import { ConnectorNameCell } from "./components/ConnectorNameCell";
+import { ConnectionTableDataItem, SortOrderEnum } from "./types";
+import { NextTable } from "../ui/NextTable";
 
 interface IProps {
```
```suggestion
interface ConnectionTableProps {
```

YatsukBogdan1 commented:
Changes (besides resolved comments):
- merged @matter-q PR for supporting responsive columns; and applied those changes on `<NextTable />`
- fixed styles for `<StremFieldsTable />` component
- removed unused `columnSort` property for `<NextTable />` component (probably will be implemented later while migrating to `react-table` sorting mechanism)

dizel852 reviewed (approved):
Did another review - LGTM 👍 
Also, tested locally, on OSS and Cloud envs - looks great. 
Seems like all the comments are fixed. 👍 

Passing to @edmundito and @tealjulia for final review

tealjulia reviewed (commented):
Is this alignment correct?  Should the checkbox be under the sync title?

<img width="191" alt="Screenshot 2023-01-31 at 12 51 33 PM" src="https://user-images.githubusercontent.com/63751206/215842720-590402ab-52dc-42cd-bc5f-6d82251223e0.png">

dizel852 commented:
> Is this alignment correct? Should the checkbox be under the sync title?
> 
> <img alt="Screenshot 2023-01-31 at 12 51 33 PM" width="191" src="https://user-images.githubusercontent.com/63751206/215842720-590402ab-52dc-42cd-bc5f-6d82251223e0.png">

Hmm, can't say for sure, but in `master` we have the same alignment. 
The Figma design is a little bit different: we don't have there a checkbox for select/unselect all streams

I've asked Bohdan to make the width of the Sync column a little narrower than we have in `master`.

tealjulia reviewed (approved):
Lgtm!  Excited to get the new components in place.
